fedimint-client/src/meta.rs

```diff
@@ -0,0 +1,432 @@
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::future::pending;
+use std::pin::pin;
+use std::time::Duration;
+
+use anyhow::{bail, Context as _};
+use async_stream::stream;
+use fedimint_core::db::{
+    Committable, Database, DatabaseTransaction, IDatabaseTransactionOpsCoreTyped,
+};
+use fedimint_core::task::waiter::Waiter;
+use fedimint_core::task::{timeout, MaybeSend, MaybeSync};
+use fedimint_core::util::{retry, FibonacciBackoff};
+use fedimint_core::{apply, async_trait_maybe_send};
+use futures::future::join_all;
+use futures::StreamExt;
+use serde::de::DeserializeOwned;
+use tokio::sync::Notify;
+use tokio_stream::Stream;
+use tracing::{debug, instrument, warn};
+
+use crate::db::{MetaFieldKey, MetaFieldName, MetaFieldPrefix, MetaFieldValue};
+use crate::Client;
+
+/// Tries to parse `str_value` as JSON. In the special case that `V` is `String`
+/// we return the raw `str_value` if JSON parsing fails. This necessary since
+/// the spec wasn't clear enough in the beginning.
+#[instrument(err)] // log on every failure
+pub fn parse_meta_value_static<V: DeserializeOwned + 'static>(
+    str_value: &str,
+) -> anyhow::Result<V> {
+    let res = serde_json::from_str(str_value)
+        .with_context(|| format!("Decoding meta field value '{str_value}' failed"));
+
+    // In the past we encoded some string fields as "just a string" without quotes,
+    // this code ensures that old meta values still parse since config is hard to
+    // change
+    if res.is_err() && std::any::TypeId::of::<V>() == std::any::TypeId::of::<String>() {
+        let string_ret = Box::new(str_value.to_owned());
+        let ret: Box<V> = unsafe {
+            // We can transmute a String to V because we know that V==String
+            std::mem::transmute(string_ret)
+        };
+        Ok(*ret)
+    } else {
+        res
+    }
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait MetaSource: MaybeSend + MaybeSync {
+    /// This must be stable.
```

Comment:
Wouldn't it be easier to give use `&'stat str`s for this?

Reply:
agreed, done